### ISSUE TITLE
[Bug]: Axes3D.computed_zorder of does not appear to work (Line3D)

### ISSUE THREAD

**Issue by mueslo** [matplotlib/matplotlib#25033]
### Bug summary

Documentation claims that artists added to Axes3D will have their zorder calculated automatically based on average location if computed_zorder is True (default). This is false (at least for Line3D).

### Code for reproduction

```python
from mpl_toolkits.mplot3d.art3d import Poly3DCollection, Line3D

import matplotlib.pyplot as plt

fig, ax = plt.subplots(subplot_kw=dict(projection='3d', computed_zorder=True))

ax.add_artist(Line3D((1,1),(0,0),(0,1), color='k'))
ax.add_artist(Line3D((1,0),(0,0),(1,0), color='b'))
ax.add_artist(Line3D((0,1),(1,1),(0,0), color='r'))

plt.show()
```

### Actual outcome

![computed_zorder](https://user-images.githubusercontent.com/847751/213581556-6f144847-7ddf-4bbb-8e30-d8328b411125.png)

Based on "average position of the [Artist](https://matplotlib.org/stable/api/artist_api.html#matplotlib.artist.Artist)s along the view direction" blue and black lines should be drawn on top of the red line. But they are obviously rendered according to the order they were added to the Axes3D.

### Expected outcome

Black on top of both red and blue.

### Matplotlib Version

3.6.3

### Installation

pip

**Comment by mueslo** [comment]
https://github.com/matplotlib/matplotlib/blob/c23ccdde6f0f8c071b09a88770e24452f2859e99/lib/mpl_toolkits/mplot3d/axes3d.py#L445-L453 It indeed only applies it to Collections and Patches. Not all Artists.

**Comment by joddeepesh-cloud** [comment]
From the code snippet shared above, it seems that computed_zorder is only applied to Collections and Patches, not to all Artist subclasses such as Line3D.
Would updating the documentation to clarify this behavior be the correct resolution for this issue?

**Comment by timhoffm** [comment]
Documentation clarifies the current state.

I'm unclear though whether this is desired or just an omission (Collections and Patches are the "area-like" artists for which proper ording is extremely important as areas can easily hide other artists. Though Lines could benefit from being drawn in the correct order as well).

OTOH it could be that lines were excluded because the simplified assumption of an "average z" is less justified. Lines typically span a large space, but cover little area, whereas patches are often more localized but cover the surrounding of the average z. It may be that average z would produce too many artefacts for lines and they are therefore left completely under user control.

This would need more investigation and probably digging into history.

Semi-OT: It may be reasonable to introduce API to control participation in zorder calculation on the artist level. Spitballing: Something like
- specify a zorder range that is used for the auto-ordering. - This allows users to explicitly place artists behind or in front by using zorders outside that range
- Artists with zorder="auto" are sorted into that range.
- Optional: Would there be a benefit of allowing multiple such range groups?

**Comment by joddeepesh-cloud** [comment]
Thanks for the clarification!
I agree that while the documentation reflects the current behavior, it’s still unclear whether excluding Line3D from computed_zorder is intentional design or an omission.
I’d be happy to investigate this further — particularly:
Looking into the history of when computed_zorder was introduced
Checking whether there were discussions around excluding line artists
Experimenting with enabling participation for Line3D to see what artefacts arise
If it turns out that this is more of a design limitation than a documentation issue, perhaps we could explore an API-level opt-in mechanism as you suggested.
Let me know if this direction makes sense — I’d be glad to dig into it.

**Comment by timhoffm** [comment]
Thinking slightly more about this, plainly Line3D to computed_zorder will not be helpul. It would fix some cases, but break others. The balance may or may not be slightly better, but the non-working cases would still be significant, not justifying the investment and the behavioral change.

A substantial improvement can only be made through artist-level configurability and ways to reliably set artists infront/behind the auto ordered ones. Maybe @scottshambaugh has thoughts on this?

**Comment by scottshambaugh** [comment]
Just reading the code so this should be double checked, but my understanding is:
The spines & ticks are plotted with zorder = 2, and each 3D artist increments +1 on top of that. So you can reliably set low zorder artists in the (2, 3) range. The upper end is unbounded, but in practice you will have a low number of artists so can manually set high zorders around a large value.

IMO all 3D artists should have the same behavior, and we can consider the ones which don't compute zorder to be a visual bug that we can fix without worrying about deprecation. I'm not clear on the history, but I would guess is that this inconsistency is just an oversight.

In general zorder is a mess, but the "2.5D" rendering of 3D data in constant zorder layers is a fundamental limitation of the rendering pipeline at the moment. Computed zorder based on average depth makes sense to me as the best default.

**Comment by timhoffm** [comment]
> my understanding is:

Yes that's likely correct. My question is: Is it sufficient to document that, or do we want some cleaner numbers/approach - e.g. reserve the range [10...100[ for auto-zorder, so that you have:
- z<10: use these values to place artists below
- 10<=z<100: auto-zorder - these values will typically not be set explicitly by users
- 100<=0: use these values to place artists above

Since z is a float the range (2,3) for "under" would work, but (i) it may be easier to rember the decades than the range (2, 3) and (ii) I find it a bit easier to work with integers 3, 4, 5 than with fractional numbers 2.1, 2.2, 2.3.

>  all 3D artists should have the same behavior [...] Computed zorder based on average depth makes sense to me as the best default

Including all by defualt is reasonable as soon as we have the ability to remove individual ones. These features are tied together. Othewise we may render the automatism unusable for certain currently working scenarios.

**Comment by mueslo** [comment]
> Just reading the code so this should be double checked, but my understanding is: The spines & ticks are plotted with zorder = 2, and each 3D artist increments +1 on top of that. So you can reliably set low zorder artists in the (2, 3) range. The upper end is unbounded, but in practice you will have a low number of artists so can manually set high zorders around a large value.
> 
> IMO all 3D artists should have the same behavior, and we can consider the ones which don't compute zorder to be a visual bug that we can fix without worrying about deprecation. I'm not clear on the history, but I would guess is that this inconsistency is just an oversight.
> 
> In general zorder is a mess, but the "2.5D" rendering of 3D data in constant zorder layers is a fundamental limitation of the rendering pipeline at the moment. Computed zorder based on average depth makes sense to me as the best default.

I agree with this, and this was also my local solution to make it it "more" 3D by patching the source code to use computed_zorder for Line3D. In my application code I then subdivided line segments until the 3D rendering was correct. This makes a lot of things easier, since using zorder and shortening lines guarantees there is some limit of subdivisions above which the 2.5D rendering will be equivalent to the 3D rendering (when only using Line3D segments). In my case it was for drawing (reciprocal) crystal lattices, i.e. 3D polygon edges.

**Comment by scottshambaugh** [comment]
At this point I think that explicitly reserving a zorder range would just add complexity with a decent likelihood of breaking people's plots where they've set manual zorders. But I'd support some guidance in the docs that better explain the current situation. We have a 2D zorder explainer (https://matplotlib.org/devdocs/gallery/misc/zorder_demo.html), but not a 3D one.

> In my application code I then subdivided line segments until the 3D rendering was correct.

Yeah I've experimented with this, it's a good way to get an 80% solution. You still have occlusion issues on the scale of individual segments, but that doesn't matter for some use cases.
